### PR TITLE
Add shadow module for debugging

### PR DESCRIPTION
**Pull request by hx89**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#30738 Add shadow module for debugging**

Differential Revision: [D18795376](https://our.internmc.facebook.com/intern/diff/D18795376/)

Add shadow module and corresponding functions for debugging.

**Motivation**

Quantization is good when it works, but difficult to know what's wrong when it doesn't satisfy the accuracy we expect. Debugging the accuracy issue of quantization is not easy and time consuming. So it's important to have a systematic debugging methodology and corresponding tools to streamline the debugging process. 

The first step of debugging is to enable the measurement of the quantization error to know how much different the quantized module is compared with float module. Currently there's no easy way to do it in PyTorch, thus we are proposing to build the shadow comparison tool to help the quantization error measurement. 

**High level design**

The basic idea of shadow comparison is that we attach float operator to quantized operator so that they take the same input, and we compare their output to compute the error.

PyTorch changes are needed for this to work, since shadow comparison needs to have both float and quantized version of the module, but during quantization the original float modules and weights are swapped with the corresponding quantized module, thus the float module is lost in this process. This swapping is done in the convert() function in quantize.py. So in order to keep the original float modules during quantization, we need to do it in the convert() function. 

We want to start with computation intensive modules such as linear and conv defined in SWAPPABLE_MODULES below, since they are frequently used and usually the main source of quantization error, and we can expand to other modules.
```
SWAPPABLE_MODULES = (nni.ConvBn2d,
                     nni.ConvBnReLU2d,
                     nni.LinearReLU,
                     nni.ConvReLU2d)
```
**How to add shadow module**

In order to add shadow module, during the quantization from float module to quantized module, instead of swapping the float model with quantized model, we want to attach float module to the quantized module so that they share the same input. And then we need to set the output of float operator to a hanging output not consumed by any other operator. Next we can attach recording observer to collect the tensor value at the output of the float and quantized operators to compute l2 error. The detailed design and considerations are explained below.

1. We first want to have a special quantize function quantize_shadow() to quantize the model we want to add shadow comparison. And quantize_shadow() function will look like below, calling transform_shadow() function instead of convert().
```
def quantize_shadow(model, run_fn, run_args, mapping=None, inplace=False):
    if mapping is None:
        mapping = DEFAULT_MODULE_MAPPING
    if not inplace:
        model = copy.deepcopy(model)
    model.eval()
    prepare(model, inplace=True)
    run_fn(model, run_args)
    transform_shadow(model, mapping, inplace=True)
    return model
```

2. In convert_shadow() function, if the type of the module is in SWAPPABLE_MODULES, instead of calling swap_module(), we call add_shadow_module() to add the shadow module as below. 
```
def transform_shadow(module, mapping=None, inplace=False):
    if mapping is None:
        mapping = DEFAULT_MODULE_MAPPING
    if not inplace:
        module = copy.deepcopy(module)
    reassign = {}
    # TODO(jerryzh): remove after deciding on the impl of intrinsic modules
    # This is required because intrinsic modules right now are implemented as
    # nn.Sequential and we don't want to swap their constituents
    SWAPPABLE_MODULES = (nni.ConvBn2d,
                         nni.ConvBnReLU2d,
                         nni.LinearReLU,
                         nni.ConvReLU2d)

    for name, mod in module.named_children():
        if type(mod) not in SWAPPABLE_MODULES:
            transform_shadow(mod, mapping, inplace=True)
        if type(mod) in SWAPPABLE_MODULES:
            reassign[name] = add_shadow_module(mod, mapping)
        else:
            reassign[name] = swap_module(mod, mapping)

    for key, value in reassign.items():
        module._modules[key] = value

    return module
```

3. Add shadow module. This is the key and challenging part. We can’t just attach a float module to the quantized module since in PyTorch eager mode we need to connect them correctly in the form of forward path. One solution proposed here is that in add_shadow_module, instead of returning the quantized module, it creates a new Shadow module which contains both quantized module and original float module.
```
def add_shadow_module(mod, mapping):
    new_mod = mod
     # Always replace dequantstub with dequantize
    if hasattr(mod, 'qconfig') and mod.qconfig is not None or type(mod) == DeQuantStub:
        if type(mod) in mapping:
            if type(mod) == QuantStub or type(mod) == DeQuantStub:
                new_mod = mapping[type(mod)].from_float(mod)
            else:
                qmod = mapping[type(mod)].from_float(mod)
                new_mod = Shadow(qmod, mod)
    return new_mod
```

The code for Shadow module is below. The Shadow module takes quantized module and float module as input, and creates a forward path inside to make the float module to shadow quantized module. The float module needs the float input, so we need to add a dequant module.
```
class Shadow(nn.Module):
    def __init__(self, q_module, float_module):
        super(Shadow, self).__init__()
        self.orig_module = q_module
        self.shadow_module = float_module
        self.dequant = nnq.DeQuantize()

        self.orig_ob = torch.quantization.RecordingObserver()
        self.shadow_ob = torch.quantization.RecordingObserver()

    def forward(self, x):
        output = self.orig_module(x)

        # Output is in float for dynamic quantization
        if output.is_quantized:
            self.orig_ob(output.dequantize())
        else:
            self.orig_ob(output)

        # Handle dynamic quantization
        if x.is_quantized:
            x = x.dequantize()

        shadow_output = self.shadow_module(x)
        self.shadow_ob(shadow_output)
        return output
```

4. Run the quantized model and collect the observer results. Once we add the recording observer, we can run the quantized model. After the run, get_observer_dict() can be used to collect all the observer into a dictionary, and we can use this dictionary to access the tensor values for each input.
```
ob_dict = get_observer_dict(model)
```
5. Compute the module level quantization error. In this step we go through the dict and find the keys corresponding to shadow module, then the module level quantization error can be computed by comparing the shadow module’s output and shadow output.
```
def report_module_shadow_error(ob_dict)
```

### CODE REVIEW

**Comment by kostmo** [comment]
## CircleCI build failures summary
As of commit b3aaaac:
* **2/2** broken upstream at merge base 409151e (see [grid view](https://dr.pytorch.org/master-timeline.html?count=1&sha1=409151e1bb90c66546901d372f59930f72242ee9&should_suppress_scheduled_builds=true&should_suppress_fully_successful_columns=true&max_columns_suppress_successful=35))
    * You may want to rebase on the `viable/strict` branch (see its [recency history](https://dr.pytorch.org/master-viable-commits.html)):
        * If your commit is newer than `viable/strict`, you can try basing on an older, stable commit:
          ```
          git fetch viable/strict
          git rebase --onto viable/strict $(git merge-base origin/master HEAD)
          ```
        * If your commit is older than `viable/strict`:
          ```
          git fetch viable/strict
          git rebase viable/strict
          ```
* **0/2** failures introduced in this PR



## Detailed failure analysis

One may explore the probable reasons each build failed interactively [on the Dr. CI website](https://dr.pytorch.org/commit-details.html?sha1=b3aaaac8278c72313547c8780767d7c95a14d168).

### 1 upstream failure recognized by patterns:
These builds matched patterns, but were probably caused by upstream breakages:
* [caffe2_onnx_py2_gcc5_ubuntu16_04_test](https://dr.pytorch.org/build-details.html?build_id=75443618)



---

<details><summary>This comment was automatically generated by <a href="https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment">Dr. CI</a> (expand for details).</summary>Follow <a href="https://dr.pytorch.org/admin/comments-opt-out.html">this link to opt-out</a> of these comments for your Pull Requests.

Please report bugs/suggestions on the <a href="https://github.com/kostmo/circleci-failure-tracker/issues">GitHub issue tracker</a>.
</details>
